File: core/document.cpp
/*
    This file is part of Rekall.
    Copyright (C) 2013-2014

    Project Manager: Clarisse Bardiot
    Development & interactive design: Guillaume Jacquemin & Guillaume Marais (http://www.buzzinglight.com)

    This file was written by Guillaume Jacquemin.

    Rekall is a free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "document.h"

Document::Document(ProjectBase *_project) :
    DocumentBase(_project) {
    project     = _project;
    project->addDocument(this);
}

Tag* Document::createTag(qint16 _versionSource, qint16 versionDest) {
    qint16 versionSource = getMetadataIndexVersion(_versionSource);
    Tag *newTag = 0;
    if(tags.count()) {
        QList<Tag*> tagsSource;
        foreach(Tag *tag, tags) {
            tag->setDocumentVersion(-1);
            if(tag->getDocumentVersion() == (versionSource-1))
                tagsSource << tag;
        }
        foreach(Tag *tagSource, tagsSource)
            createTag(tagSource, tagSource->getDocumentVersion()+1);
    }
    else {
        qreal tS = Global::alea(5, 120);
        if(getFunction() == DocumentFunctionRender)
            tS = Global::alea(0, 5);
        newTag = createTag(TagTypeContextualTime, tS, -1, versionDest);
    }
    return newTag;
}
Tag* Document::createTag(TagType type, qreal timeStart, qreal duration, qint16 version) {
    Tag *newTag = new Tag(this, version);
    newTag->init(type, timeStart, duration);
    tags.append(newTag);
    return newTag;
}
Tag* Document::createTag(Tag *tagSource, qint16 version) {
    if(version < -1)
        version = tagSource->getDocumentVersionRaw();
    Tag *newTag = createTag(tagSource->getType(), tagSource->getTimeStart(), tagSource->getDuration(), version);
    newTag->setTimeMediaOffset(tagSource->getTimeMediaOffset());
    newTag->timelinePos     = tagSource->timelinePos;
    newTag->timelineDestPos = tagSource->timelineDestPos;
    newTag->viewerPos       = tagSource->viewerPos;
    newTag->viewerDestPos   = tagSource->viewerDestPos;
    return newTag;
}
void Document::removeTag(void *tag) {
    removeTag((Tag*)tag);
}
void Document::removeTag(Tag *tag) {
    tags.removeOne(tag);
    //delete tag;
}

QDomElement Document::serialize(QDomDocument &xmlDoc) const {
    QDomElement xmlData = xmlDoc.createElement("document");
    if(tags.count()) {
        QDomElement tagsXml = xmlDoc.createElement("tags");
        foreach(Tag *tag, tags)
            tagsXml.appendChild(tag->serialize(xmlDoc));
        xmlData.appendChild(tagsXml);
    }
    xmlData.appendChild(serializeMetadata(xmlDoc));
    return xmlData;
}
void Document::deserialize(const QDomElement &xmlElement) {
    deserializeMetadata(xmlElement);
    QDomNode tagsNode = xmlElement.firstChild();
    while(!tagsNode.isNull()) {
        QDomElement tagsElement = tagsNode.toElement();
        if((!tagsElement.isNull()) && (tagsElement.nodeName() == "tags")) {
            QDomNode tagNode = tagsElement.firstChild();
            while(!tagNode.isNull()) {
                QDomElement tagElement = tagNode.toElement();
                if((!tagElement.isNull()) && (tagElement.nodeName() == "tag")) {
                    qreal timeStart = tagElement.attribute("timeStart").toDouble();
                    qreal timeEnd   = tagElement.attribute("timeEnd").toDouble();
                    createTag((TagType)tagElement.attribute("type").toInt(), timeStart, timeEnd-timeStart, tagElement.attribute("documentVersion").toInt());
                }
                tagNode = tagNode.nextSibling();
            }
        }
        tagsNode = tagsNode.nextSibling();
    }
}

File: http/www/js/rekall/Document.js
/*
    This file is part of Rekall.
    Copyright (C) 2013-2014

    Project Manager: Clarisse Bardiot
    Development & interactive design: Guillaume Jacquemin & Guillaume Marais (http://www.buzzinglight.com)

    This file was written by Guillaume Jacquemin.

    Rekall is a free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

function Document() {
	this.tags      		   = new Array();
	this.metadatas		   = new Array();
	this.currentVersion    = 0;
	this.goodVersion       = 0;
}
Document.prototype.addTag = function(tag) {
	this.tags.push(tag);
}

Document.prototype.initMetadatas = function(version) {
	if(version == undefined)
		version = this.currentVersion;
	if(this.metadatas[version] == undefined)
		this.metadatas[version] = new Metadata();
	return version;
}
Document.prototype.getMetadatas = function(version) {
	version = this.initMetadatas(version);
	return this.metadatas[version];
}
Document.prototype.setMetadatas = function(metadatas, version) {
	version = this.initMetadatas();
	this.metadatas[version] = metadatas;
}
Document.prototype.isVideoOrAudio = function(version) {
	return this.isAudio(version) || this.isVideo(version); 
}
Document.prototype.isAudio = function(version) {
	return this.getMetadata("File->MIME Type", version).startsWith("audio"); 
}
Document.prototype.isVideo = function(version) {
	return this.getMetadata("File->MIME Type", version).startsWith("video"); 
}
Document.prototype.isMarker = function(version) {
	return this.getMetadata("File->MIME Type", version) == "Marker"; 
}
Document.prototype.isImage = function(version) {
	return this.getMetadata("File->MIME Type", version).startsWith("image"); 
}


Document.availableMetadataKeys = new Object();
Document.prototype.getMetadata = function(metadataKey, version) {
	if(metadataKey == "")
		return "";
	if(this.getMetadatas(version)[metadataKey] != undefined)
		return this.getMetadatas(version)[metadataKey];
	else
		return undefined;
}
Document.prototype.setMetadata = function(metadataKey, metadataValue, version) {
	var currentValue = "";
	if(metadataKey instanceof Metadata) {
		metadataValue = metadataKey.content + "";
		metadataKey   = metadataKey.metadataKey + "";
	}
	if((metadataKey.toLowerCase() != "exiftool") && (metadataValue.toLowerCase().indexOf("use -b option to extract") === -1)) {
		currentValue = this.getMetadatas(version)[metadataKey];
		this.getMetadatas(version)[metadataKey] = metadataValue;
	
		if(Document.availableMetadataKeys[metadataKey] == undefined)
			Document.availableMetadataKeys[metadataKey] = {count: 0};
		Document.availableMetadataKeys[metadataKey].count++;
	}
	return (currentValue != metadataValue);
}

Document.prototype.compareMetadatas = function(metadatas) {
	if(metadatas == undefined)
		metadatas = this.getMetadatas(-1);
		
	var metadatasCompared = new Metadata();
	var metadatasSource   = this.getMetadatas();
	var thiss = this;
	for (var metadataKey in metadatas) {
		var metadataValue = metadatas[metadataKey];
		if(metadatasSource[metadataKey] != metadataValue)
			metadatasCompared[metadataKey] = metadataValue;
	}
	return metadatasCompared;
}
Document.prototype.cloneMetadatas = function(version) {
	var metadatasCloned = new Metadata();
	for (var metadataKey in this.getMetadatas(version))
		metadatasCloned[metadataKey] = this.getMetadatas(version)[metadataKey];
	return metadatasCloned;
}


Document.prototype.addVersion = function(document) {
	var thisVersion = this.currentVersion;
	this.currentVersion++;
	this.goodVersion++;
	this.setMetadatas(document.getMetadatas(), this.currentVersion);
	var thiss = this;
	if(document.tags.length == 0) {
		for (var index in this.tags) {
			var tag = this.tags[index];
			if(tag.version == thisVersion) {
				var newTag = new Tag(thiss);
				newTag.timeStart = tag.getTimeStart();
				newTag.timeEnd   = tag.getTimeEnd();
				thiss.addTag(newTag);
			}
		}
	}
	else {
		for (var index in document.tags) {
			var tag = document.tags[index];
			tag.version = thiss.currentVersion;
			tag.document = thiss;
			thiss.tags.push(tag);
		}
	}
}



Document.prototype.openFile = function() {
	$.ajax(Utils.getLocalFilePath(this, "open"));
}
Document.prototype.openFinder = function() {
	$.ajax(Utils.getLocalFilePath(this, "finder"));
}
Document.prototype.openBrowser = function() {
	window.open(Utils.getLocalFilePath(this, "file"), "_blank");
}
Document.prototype.downloadFile = function() {
	window.open(Utils.getLocalFilePath(this, "download"), "_blank");
}

function Metadata() {
}